### PR TITLE
Add support for function widgets

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
Following on from #1816, particularly discussion here https://github.com/napari/napari/pull/1816#issuecomment-725680671 this PR adds the ability to provide functions that we turn into dock widgets using magic gui via an `viewer.window.add_magic_function(function, magic, dock)` API, where `function` is the function, `magic` is a dict that gets passed to the `magicgui` method and `dock` is a dict that gets passed to `add_dock_widget` method.

This PR adds two examples `magic_image_arthimetic` and `magic_parameter_sweep` based on respective examples of in the magicgui docs https://magicgui.readthedocs.io/en/latest/

I've made a couple issues on the magic gui around being able to use strings for types, see https://github.com/napari/magicgui/issues/46 and https://github.com/napari/magicgui/issues/47. I'm also curious about the ramifications of the API changes in https://github.com/napari/magicgui/pull/43, but we don't necessarily need to wait to release this until that is merged, though it might be nice.

Once we have this merged it will be much easier to continue with #1816 (or supercede it) as now a plugin will just be able to provide a list of tuples of inputs to `add_magic_function`.

I havn't added tests yet, but will do so. I think it would be nice to get some review and input on this first though - cc @tlambert03 @GenevieveBuckley @jni 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: examples/magic_image_arithmetic.py
Comment (GenevieveBuckley) [inline examples/magic_image_arithmetic.py:14]: Silly question maybe, but what happens if a user tries diving by an image containing zeros? Does it get handled nicely or will that cause a problem?

File: examples/magic_image_arithmetic.py
Comment (GenevieveBuckley) [inline examples/magic_image_arithmetic.py]: This comment isn't particularly useful, mostly it seems to repeat the content in the line of code below it. It would be better if it explained what the `call_button` actually is/does (could link to docs?), or if the comment was removed.

File: examples/magic_image_arithmetic.py
Comment (GenevieveBuckley) [inline examples/magic_image_arithmetic.py:1]: I think including the arithmetic and parameter sweep examples in the napari docs makes them much more discoverable to new users. I'm +1 on that!

File: examples/magic_parameter_sweep.py
Comment (GenevieveBuckley) [inline examples/magic_parameter_sweep.py:4]: It would be good to tell users what the advantages of using a custom class here are, and in which circumstances it's useful to do.

File: examples/magic_image_arithmetic.py
Comment (tlambert03) [reply examples/magic_image_arithmetic.py:14]: it would show a zero division `RuntimeWarning`, those pixels would be `inf` and napari would show them as white.  but otherwise no problem I think.  (course, the solution would be to make a smarter `image_arithmetic` function below)

File: examples/magic_parameter_sweep.py
Comment (tlambert03) [reply examples/magic_parameter_sweep.py:4]: Yeah, I'd probably even "downplay" that point. After https://github.com/napari/magicgui/pull/43, you won't be passing actual Qt widgets anyway... but rather using something like [`widget_type = 'slider'`](
https://github.com/napari/magicgui/pull/43/files#diff-2ba78d75eb081db3f919d0551038968cb823d57b7828fafa719a88105aaee634R20) ... custom widgets will still be possible, but the use case will be even less common.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]:
```suggestion
        dock.setdefault('name', function.__name__.replace("_", " "))
```

(note also that line 615 here wouldn't actually replace the `function_name` variable...)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: I think this would be a safe instance to use mutable default arguments `magic={}, dock={}`, and remove the `if dock/magic is None: dock/magic = {}` ... (at least, I can't think of a scenario with this code where that would lead to unexpected results, @jni?)

then the whole thing would just be:
```python
widget = magicgui(**magic)(function).Gui()
self.qt_viewer.viewer.layers.events.changed.connect(
    lambda x: widget.refresh_choices()
)
dock.setdefault('name', function.__name__.replace("_", " "))
return self.add_dock_widget(widget, **dock)
```

Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: also...
- I wonder about naming it `add_function_widget` rather than `add_magic_function`?  Says a little more about _what_ it does rather than _how_ it does it.  (but don't feel strongly)

- I'd like to hear people's thoughts on making `magic` and `dock` just `**kwargs`... and then pulling out the appropriate arguments for each internal function.  would allow 
```python
add_function_widget(func, area='right')
# instead of 
add_function_widget(func, dock={'area': 'right'})

# and
add_function_widget(func, auto_call=True)
# instead of
add_function_widget(func, magic={'auto_call': True})
```

**Review by tlambert03 — COMMENTED** [review]
this looks great to me @sofroniewn, exactly how I pictured it.  I made a couple small implementation comments.  but don't feel strongly on any of them


> I've made a couple issues on the magic gui around being able to use strings for types, see napari/magicgui#46 and napari/magicgui#47. 

totally agree.  will do.


> I'm also curious about the ramifications of the API changes in napari/magicgui#43,

The main change as it relates to this PR would be around the `.Gui()` syntax (which was always a bit weird).  Briefly, the new `magicgui` decorator will return an instance of `magicgui.FunctionGui`, which _itself_ is a magicgui widget (rather than simply adding a function attribute `Gui` that points to the widget).  outside of napari, that object could be shown just with `decorated_function.show()` (or `decorated_function.show(run=True)` to spin up an event loop).  To get at the actual `QWidget`, you would now use `.native`, much like how vispy does it.  so, in this PR, you'd change it to
```
widget = magicgui(**magic)(function).native
```

but otherwise, it shouldn't change much here.  To get a more complete idea about what is changing from the API perspective, have a look at how the "examples" are changing in https://github.com/napari/magicgui/pull/43/files

**Comment by tlambert03** [comment]
oh... also, after https://github.com/napari/magicgui/pull/43/

`viewer.layers.events.changed.connect(lambda x: widget.refresh_choices())`
will become
`viewer.layers.events.changed.connect(widget.reset_choices)`

**Comment by sofroniewn** [comment]
> oh... also, after napari/magicgui#43
>
> viewer.layers.events.changed.connect(lambda x: widget.refresh_choices())
> will become
> viewer.layers.events.changed.connect(widget.reset_choices)

will each widget always have a `refresh_choices()` or `reset_choices` on it? i.e. even if there is no `layer` argument passed? Does it do a global refresh of all enum / list etc type inputs?

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: I'm open to both changes, curious what @jni and others think

File: examples/magic_parameter_sweep.py
Comment (sofroniewn) [reply examples/magic_parameter_sweep.py:4]: I might be inclined to change this example then until after the next magicgui release, maybe even hold off on this example entirely, as adding sliders will be a big thing people want to do and I'm not sure we want to put out a more complex way to do it that we know will change to a simpler one

**Comment by tlambert03** [comment]
> will each widget always have a `refresh_choices()` or `reset_choices` on it? i.e. even if there is no `layer` argument passed? 

not all widgets, but all [`CategoricalWidgets`](https://github.com/tlambert03/magicgui/blob/interface/magicgui/widgets.py#L316) ... (a widget will be categorical if it is an `Enum` or was provided with `choices` parameter.)

> Does it do a global refresh of all enum / list etc type inputs?

it depends on how the categorical widget was created. In this case though, a parameter annotated as a napari `Layer` will have `choices` set to the `get_layers` function (in [`napari.utils._magicgui`](https://github.com/napari/napari/blob/master/napari/utils/_magicgui.py#L41)).  So reset_choices will basically just call that function (which will grab layers in the layerlist)

File: examples/magic_parameter_sweep.py
Comment (tlambert03) [reply examples/magic_parameter_sweep.py:4]: I agree.  One of the most common "mis-uses" of magicgui I've seen is people providing `widget_type: QSpinBox` ... even if the parameter is an `int` ... which makes me die a little inside :) cause the whole point was to not have to use QtWidgets directly.  And I think it's probably because of this slider example... so yeah, I'd hate to see that pattern proliferate.

**Comment by sofroniewn** [comment]
> not all widgets, but all CategoricalWidgets ... (a widget will be categorical if it is an Enum or was provided with choices parameter.)

Do I then need to add a check to see if the function/ attribute exists or is it always safe to try and connect in the way I'm doing?

**Comment by tlambert03** [comment]
> > not all widgets, but all CategoricalWidgets ... (a widget will be categorical if it is an Enum or was provided with choices parameter.)
> 
> Do I then need to add a check to see if the function/ attribute exists or is it always safe to try and connect in the way I'm doing?

ohhh, right!  good point :) yeah, I think we'll need to check.  but actually, this concept probably belongs in `napari.utils._magicgui` more than it does here.  in any case, let's just remember to revisit this after the new magicgui

**Comment by sofroniewn** [comment]
> ohhh, right! good point :) yeah, I think we'll need to check. but actually, this concept probably belongs in napari.utils._magicgui more than it does here. in any case, let's just remember to revisit this after the new magicgui

For now I'll check and then make an issue to follow-up

**Comment by tlambert03** [comment]
> For now I'll check and then make an issue to follow-up

until the magicgui API changes, all widgets _will_ have a [`refresh_choices` method
](https://github.com/napari/magicgui/blob/a1385820ca25466325e3a3771114bcef75c2e39f/magicgui/core.py#L413)... and I believe it should be safe to call it even if none of the subwidgets have registered choices (it will just do nothing).

File: napari/_qt/qt_main_window.py
Comment (jni) [reply napari/_qt/qt_main_window.py]: > (at least, I can't think of a scenario with this code where that would lead to unexpected results, @jni?)

Reading the body of the function, I agree, but literally yesterday, for a tiny utility function in an interactive session, for the same issue, I was like, "nah, I don't need to worry about it here" and sure enough it exploded in my face. 😂 So I think we should stick to accepted good practice and keep the Nones, even if it's a bit uglier.

I *can* imagine future contributors coming in here and messing with the dictionary and breaking things unexpectedly.

> I wonder about naming it add_function_widget rather than add_magic_function? Says a little more about what it does rather than how it does it. (but don't feel strongly)

Haha "how it does it"... "with magic!" 😂 I like `add_function_widget` quite a bit more than `add_magic_function`. I presume we envision this working with both magical and muggle functions. 😜 

> I'd like to hear people's thoughts on making magic and dock just **kwargs... and then pulling out the appropriate arguments for each internal function. would allow

One issue is whether we envision magicgui and dock kwargs ever clashing. It's not very far-fetched, is it? One suggestion: do both? Allow `**kwargs` with smart unpacking, *and* add two parameters, `magicgui_kwargs` and `dock_kwargs`, where people who like being explicit, or, in the case of clashes, people who *need* to be explicit, can do so.

File: examples/magic_image_arithmetic.py
Comment (jni) [reply examples/magic_image_arithmetic.py:14]: Well, 0/0 will give nan and then I think vispy becomes unhappy. See vispy/vispy#1509 and vispy/vispy#1511 for unfinished solutions to this problem.

**Review by jni — APPROVED** [review]
I agree with almost everything @tlambert03 wrote. The broad strokes here though are super great, and I think @sofroniewn will know how to wrap this up, so I'm providing my approval pending the discussed changes.

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: > One issue is whether we envision magicgui and dock kwargs ever clashing. It's not very far-fetched, is it? One suggestion: do both? Allow **kwargs with smart unpacking, and add two parameters, magicgui_kwargs and dock_kwargs, where people who like being explicit, or, in the case of clashes, people who need to be explicit, can do so.

hmm, I'm a little unsure about opening up both, if what about just doing one, so `magic` was the magicgui dict, which is maybe more involved, and then there was no `dock` keyword arg which is maybe a bit weird, but we just had the same kwargs as the `add_dock_widget`?

I think that might be quite nice actually. The API for `add_function_widget` and `add_dock_widget` are now basically the same apart from the function one takes a function instead of a widget directly, and the function one takes an additional kwarg `magic`.

I feel like if we start here and people want to just pass magic gui kwargs directly we could add that in the future. Thoughts @tlambert03 @jni?

Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: > I feel like if we start here and people want to just pass magic gui kwargs directly we could add that in the future. Thoughts @tlambert03 @jni?

I think what we do now is probably how it's "going to be" (once examples get out there, they don't tend to change easily)... :) but as I said, I don't feel too strongly about it either way.  I'm also ok with your proposal for `magic` or `magic_kwargs` being a dict, and then having all the standard dock_widget kwargs being spelled out.  your call

Comment (jni) [reply napari/_qt/qt_main_window.py]: Dunno... I think I'd be more interested in passing magicgui kwargs, since they are harder (impossible?) to modify after instantiation... Whereas moving a dockwidget around is fine.

I think the least controversial choice here is the two kwarg dicts. (Though I do think they should have the word `kwargs` in the argument name — that is standard.) I suggest we start with that and open a followup issue for this discussion.

**Comment by sofroniewn** [comment]
> I agree with almost everything @tlambert03 wrote. The broad strokes here though are super great, and I think @sofroniewn will know how to wrap this up, so I'm providing my approval pending the discussed changes.

Ok great sounds like we're pretty close. Out of curiosity @tlambert03 do you have guestimate on a timeline for how long it would take to finish https://github.com/napari/magicgui/pull/43? I just looked and we don't actually depend on magicgui yet - I need to add that to this PR - and if possible it might be nice to make our first dependency on a version after that PR has merged, but I don't think that's a requirement.

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [reply napari/_qt/qt_main_window.py]: actually... I am swayed by @jni's point about potential clashes, particularly in the magic_gui args (which, if I'm not mistaken, can be scarily flexible since they can be named after parameter args)... so let's _not_ do what I originally suggested for magicgui **kwargs.  I'm fine with `dock_kwargs` and `magic_kwargs`.

> they are harder (impossible?) to modify after instantiation

fwiw: after the "mythical" magicgui rewrite that will merge someday... modifying magic kwargs after instantiation is quite straightforward (though the range of what's possible is going to be restricted slightly at first)

**Comment by tlambert03** [comment]
> do you have guestimate on a timeline for how long it would take to finish napari/magicgui#43?

the code is mostly done, (just need to add back labels).  The main issues now are tests, and updated docs... and that's unfortunately why it's sat there for a while.  but I can probably just live with imperfect test coverage for the first . probably a week or two of focused work?  (question is when will the focused work come :joy:)

**Comment by sofroniewn** [comment]
> the code is mostly done, (just need to add back labels). The main issues now are tests, and updated docs... and that's unfortunately why it's sat there for a while. but I can probably just live with imperfect test coverage for the first . probably a week or two of focused work? (question is when will the focused work come 😂)

ok sounds good, we can play it by ear!

**Comment by jni** [comment]
Given https://github.com/napari/magicgui/pull/43 and how much that will change the API, I think we should probably hold off on this until 0.4.2, despite (or perhaps because of!) the massive community excitement about this. In short, I think many people would start using this immediately, and it would become a major backwards-compatibility headache. Let's take a couple of more weeks and get this right? (Or at least, a little more right. 😂)

**Comment by tlambert03** [comment]
> Let's take a couple of more weeks and get this right? (Or at least, a little more right. 😂)

yeah, I think that's a good idea.  While the decorator syntax itself won't change too dramatically, the fact that accessing the widget with `.Gui()` will likely go away (provided everyone likes that?) is probably reason enough to wait.  I promise to get on it soon... this week is really busy, but I will try make it a priority.  (and wouldn't mind a couple extra reviews of napari/magicgui#43 ... even just casual perusals)

edit: jeez, I'm realizing now that even the original post there is a bit out of date.  so if you want to maximize your review efforts, maybe hold off on that too for a week or so

**Comment by GenevieveBuckley** [comment]
> Let's take a couple of more weeks and get this right? (Or at least, a little more right. 😂)

Agreed

File: examples/magic_image_arithmetic.py
Comment (sofroniewn) [reply examples/magic_image_arithmetic.py]: expanded comment

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: not using dict anymore, so changed

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: have renamed `add_function_widget` and gone with `magic_kwargs` and then individual kwargs for dock widget. I think there's quite nice the symmetry now between the `add_dock_widget` and `add_function_widget` API.

**Comment by sofroniewn** [comment]
Ok, I've updated this PR some, including doing some renaming and adding a test. I think it is now as close as it can be to merge until the next magicgui release at which point we can bump the version number, remove the private imports from the example, and change the connection to refresh. If there is anything else that anyone would like to see either now or then I can add that too. I think it would be really nice to get this in for the 0.4.3 release, so next ~2 weeks if possible.

**Comment by tlambert03** [comment]
was just working on magicgui now...  so it's a high priority.  but there's a fair amount of API breakage that I need to go in and deprecate/fix.  but it's on the radar...

File: examples/magic_image_arithmetic.py
Comment (sofroniewn) [inline examples/magic_image_arithmetic.py]: I'm trying this again after the merge of magicGui PR43, and if I use strings for types here instead of the classes I get the following error (I'm not sure if that is expected or not, so apologies if this is expected)

```python
(base) czirwc1macos2701:napari nsofroniew$ python examples/magic_image_arithmetic.py
ERROR:root:Unhandled exception:
Traceback (most recent call last):
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/event_loop.py", line 79, in gui_qt
    yield app
  File "examples/magic_image_arithmetic.py", line 34, in <module>
    viewer.window.add_function_widget(image_arithmetic, magic_kwargs=magic)
  File "/Users/nsofroniew/GitHub/napari/napari/_qt/qt_main_window.py", line 654, in add_function_widget
    widget = magicgui(**magic_kwargs)(function)
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/function_gui.py", line 288, in inner_func
    app=app,
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/function_gui.py", line 79, in __init__
    widgets=list(sig.widgets(app).values()),
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/signature.py", line 220, in widgets
    {n: p.to_widget(app) for n, p in self.parameters.items()}
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/signature.py", line 220, in <dictcomp>
    {n: p.to_widget(app) for n, p in self.parameters.items()}
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/signature.py", line 144, in to_widget
    options=options,
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/widgets/_bases.py", line 229, in create
    wdg_class, opts = get_widget_class(default, annotation, options)
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/type_map.py", line 156, in get_widget_class
    widget_type, _options = pick_widget_type(value, annotation, _options)
  File "/Users/nsofroniew/GitHub/magicgui/magicgui/type_map.py", line 131, in pick_widget_type
    raise ValueError(f"Could not pick widget for type: {dtype!r}")
ValueError: Could not pick widget for type: ForwardRef('napari.layers.Image')
```

I wonder if that is something that we need to fix on the magicgui side or on the napari side. This was raised initially https://github.com/napari/magicgui/issues/46 which I will acknowledge is still open

File: examples/magic_parameter_sweep.py
Comment (sofroniewn) [reply examples/magic_parameter_sweep.py:4]: This has been changed with the merge of https://github.com/napari/magicgui/pull/43

File: examples/magic_image_arithmetic.py
Comment (tlambert03) [reply examples/magic_image_arithmetic.py]: Still need to add that on the magicgui side, your napari/magicgui#46 issue is still very much valid!  Shouldn't be hard now though, hopefully just a few lines.  Will do it soon

**Comment by sofroniewn** [comment]
I've updated this to use the latest magicGui syntax and can confirm that everything works nicely. I still need to add some tests, and it might be nice if we fixed https://github.com/napari/magicgui/issues/46 before it's not required. 

Curious if anyone else has any additional ideas for this PR?

One thing I was thinking about is if we used an approach like our [ReadOnlyWrapper](https://github.com/napari/napari/blob/936d2890410ac2dc92a42e397bfd7daec37657ea/napari/utils/interactions.py#L8-L19) to make napari objects (layers and viewer) we pass read only. If I understand that correctly it won't introduce any additional time and would make our "functional" plugins a lot more functional as we'd avoid side effects. Note here one could still modify the array data in place, but even that could be prevented, like by wrapping arrays with dask. We could also thing of ways to block adding keybindings/ mouse functions, which I don't think should be possible via this API.

**Comment by sofroniewn** [comment]
>  I still need to add some tests,

ah turned out there already was one. it's failing now but probably because i'm using magicgui 0.2.0 syntax and testing against 0.1.6. that will get fixed after the magicgui release

File: examples/magic_image_arithmetic.py
Comment (tlambert03) [reply examples/magic_image_arithmetic.py]: added in https://github.com/napari/magicgui/pull/52

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: renamed and addressed

File: examples/magic_image_arithmetic.py
Comment (sofroniewn) [reply examples/magic_image_arithmetic.py]: I'n still getting an error on this, but it looks like it is on the "napari" side with something in `utils._magicgui` file

```python-traceback
(base) czirwc1macos2701:napari nsofroniew$ python examples/magic_parameter_sweep.py
/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py:217: FutureWarning:

FloatSlider.__init__() got unexpected keyword arguments {'maximum'}.
In the future this will raise an exception

  FutureWarning,
/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py:722: FutureWarning:

As of magicgui 0.2.0, a `choices` callable may accept only a single positional
argument (an instance of `magicgui.widgets.CategoricalWidget`), and must return
an iterable (the choices to show). Function 'get_layers' accepts 2 arguments.
In the future, this will raise an exception.

  FutureWarning,
ERROR:root:Unhandled exception:
Traceback (most recent call last):
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py", line 710, in choices
    _choices = choices(self)
  File "/Users/nsofroniew/GitHub/napari/napari/utils/_magicgui.py", line 130, in get_layers
    layer for layer in viewer.layers if isinstance(layer, layer_type)
  File "/Users/nsofroniew/GitHub/napari/napari/utils/_magicgui.py", line 130, in <genexpr>
    layer for layer in viewer.layers if isinstance(layer, layer_type)
TypeError: isinstance() arg 2 must be a type or tuple of types

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py", line 238, in <lambda>
    lambda *x: self.parent_changed(value=self.parent)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/events.py", line 603, in __call__
    self._invoke_callback(cb, event)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/events.py", line 625, in _invoke_callback
    cb_event=(cb, event),
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/events.py", line 619, in _invoke_callback
    cb(event)
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py", line 686, in reset_choices
    self.choices = self._default_choices
  File "/Users/nsofroniew/opt/anaconda3/lib/python3.7/site-packages/magicgui/widgets/_bases.py", line 725, in choices
    _choices = choices(self.native, self.annotation)  # type: ignore
  File "/Users/nsofroniew/GitHub/napari/napari/utils/_magicgui.py", line 130, in get_layers
    layer for layer in viewer.layers if isinstance(layer, layer_type)
  File "/Users/nsofroniew/GitHub/napari/napari/utils/_magicgui.py", line 130, in <genexpr>
    layer for layer in viewer.layers if isinstance(layer, layer_type)
TypeError: isinstance() arg 2 must be a type or tuple of types
```

**Comment by sofroniewn** [comment]
Ok all, this is ready for another look. I've added our requirement for `magicgui > 0.2.0`. I'm still having problems using strings instead of `napari.layers.Image` for types, see https://github.com/napari/napari/pull/1856#discussion_r549041360, but I think that might be now on the napari send. Would love to get feedback from @tlambert03 on this before merge. Very exciting to get this in!!

**Comment by tlambert03** [comment]
> I'm still having problems using strings instead of `napari.layers.Image`

ahh... this is because magicgui resolves the forward reference as needed, but doesn't actually _replace_ the type annotation.  It would be easy enough to fix on the magicgui side (see https://github.com/napari/magicgui/issues/65).  Otherwise, we'd also need to resolve them in `utils._magicgui`,  you could use the magicgui function for it like this:

```python
def get_layers(...):
    if layer_type is None:
        gui, layer_type = gui.native, gui.annotation

    # resolve the ref
    if isinstance(layer_type, ForwardRef):
        from magicgui.type_map import _evaluate_forwardref
        layer_type = _evaluate_forwardref(layer_type)
```

**Comment by tlambert03** [comment]
I think it's probably for the best to do this on the magicgui side, I made a PR: https://github.com/napari/magicgui/pull/66

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]:
```suggestion
            Keyword arguments to :func:`magicgui.magicgui` that
```

(this would provide a direct link if we add magicgui intersphinx to our docs)

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: here's a thought: we could modify `add_dock_widget` to connect the `layers.events` to `widget.reset_choices` when widget is an instance of `magicgui.widgets.Widget`.  That would let us remove that whole "Keep the dropdown menus in the widget in sync" line from _all_ of the examples that use it... and with a little one-liner magic, would reduce this whole function to:

```suggestion
        from magicgui import magicgui

        return self.add_dock_widget(
            magicgui(function, **magic_kwargs or {}),
            name=name or function.__name__.replace('_', ' '),
            area=area,
            allowed_areas=allowed_areas,
            shortcut=shortcut,
        )
```

**Review by tlambert03 — APPROVED** [review]
This looks great @sofroniewn!  The examples definitely make it look even easier now to get some gui goodness into napari. :)

I left some thoughts, but you can take them or leave them.  Also, if you want, you can remove the comments from `utils._magicgui.get_layers` if we're going to depend on `magicgui>=0.2.0`

**Comment by tlambert03** [comment]
also: if you want, I try to get napari/magicgui#63, napari/magicgui#64, and napari/magicgui#65 in and release 0.2.1 tomorrow and you can pin to that instead?

File: napari/_qt/qt_main_window.py
Comment (sofroniewn) [reply napari/_qt/qt_main_window.py]: makes sense to me, will do tomorrow

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]: one more suggestion: when not using as a decorator, this syntax is probably clearer:
```suggestion
            magicgui(function, **magic_kwargs or {}),
```

**Comment by sofroniewn** [comment]
I've moved the auto-connection to the `add_dock_widget` which is nice as any magicgui added widget will now benefit from it. I did it by looking for `reset_choices`, let me know @tlambert03 if you have another way in mind (what I like about this is that it doesn't mean we have to import magicgui, but we could do different too, feel free to make a suggested change if you want.

If you want to make a commit to this PR that cleans up `utils._magicgui` now that we have > 0.2.0 that would be great too, might be safer that you do it rather than me!

> also: if you want, I try to get napari/magicgui#63, napari/magicgui#64, and napari/magicgui#65 in and release 0.2.1 tomorrow and you can pin to that instead?

Sounds good, no rush

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py]:
```suggestion
            # `CategoricalWidget`s
```

File: napari/_qt/qt_main_window.py
Comment (tlambert03) [inline napari/_qt/qt_main_window.py:587]: I'm forgetting at the moment... is `layers.events` the same as "any event" or did we need to do `layers.events.inserted` and `layers.events.removed`?

Comment (sofroniewn) [reply napari/_qt/qt_main_window.py:587]: `layers.events` will do all of them

**Comment by tlambert03** [comment]
> If you want to make a commit to this PR that cleans up utils._magicgui now that we have > 0.2.0 that would be great too, might be safer that you do it rather than me!

ok, I just cleaned up `magicgui._utils.get_layers`, and I extended your test a bit to make sure that `viewer.layers` does indeed stay in sync with the layer choices in the widget.

I figured out what was going wrong in #2044 and submitted fix in #2046.
On the magicgui side, I tweaked a few more things, so if you have time to peek at and approve napari/magicgui#67, napari/magicgui#68, napari/magicgui#69, I can release 0.2.1 and we can pin to that instead.

**Comment by sofroniewn** [comment]
> On the magicgui side, I tweaked a few more things, so if you have time to peek at and approve napari/magicgui#67, napari/magicgui#68, napari/magicgui#69, I can release 0.2.1 and we can pin to that instead.

I've approved all 3 PRs, they all look good to me. If you want to make a magicgui `0.2.1rc0` tomorrow I can give that a test before you make the actual release, and then once released I'll update this PR.

**Comment by sofroniewn** [comment]
Ok this has been updated to now depend on the latest magicgui `0.2.1` and fixed so all tests should be passing. I also added a comment about forward references and link to https://www.python.org/dev/peps/pep-0484/#forward-references in the examples. @tlambert03 can you look those over (occurs in both examples), feel free to directly edit PR. If you give this one final thumbs up I will then merge!! So exciting to get this in!!!

**Comment by tlambert03** [comment]
> @tlambert03 can you look those over (occurs in both examples)

good idea.  I made a slight update to explain that ForwardRefs are mostly necessary if you don't want to import napari.  all looks good to me i think! :tada:

**Comment by sofroniewn** [comment]
Ok, great, will merge now!!

**Pull request merged**
